1st draft.

CoitThomas opened the pull request.

CoitThomas committed:
Initialize branch wip (work in progress).

CoitThomas committed:
First draft of create_database.py.

CoitThomas committed:
First draft of find_user module.
Remove transpose() function from create_database module.
Change date storage format to mm/dd/yyyy.
Remove parse_date() function from create_database module.

CoitThomas committed:
First draft of find_user_by_nationality.
Refactor shared code out into the date_validation module and parse module.

CoitThomas committed:
Add unit tests.
Remove case-sensitivity from entire program.
Include username and nationality in find_user.py final print out.
Refactor the main and namesake functions in the find_user module and the find_user_by_nationality module.

austinwiltshire commented on create_database.py:
```diff
@@ -0,0 +1,107 @@
+    assert data_lines, "There is no data!"
+    assert int(size), "An integer in the form of a string must be provided."
+    assert 0 < int(size) <= len(data_lines), "A valid size must be provided."
+    assert len(data_lines) % size == 0, "Data is incomplete."
```
great assertions

austinwiltshire commented on create_database.py:
```diff
@@ -0,0 +1,107 @@
+    assert len(data_lines) % size == 0, "Data is incomplete."
+    return [data_lines[pos:pos + size] for pos in range(0, len(data_lines), size)]
+
+def create_user_table(db_filename, users_filename):
```
Function is needlessly coupling file reading with database opening.

Why not have the function take in an active database object and an active file rather than opening them both.

austinwiltshire commented on create_database.py:
```diff
@@ -0,0 +1,107 @@
+    file.
+    """
+    # Open and connect to the database.
+    database = sqlite3.connect(db_filename)
```
Violating DRY here - these lines are the same as 31 through 32. If these functions took in an active database object rather than filenames, they could share this code.

austinwiltshire commented on create_database.py:
```diff
@@ -0,0 +1,107 @@
+    assert data, "There is no data in file: %s" % users_filename
+    data_chunks = chunk_data(data, 3)
+    # Insert data into the users table.
+    for chunk in data_chunks:
```
might want to split reading and parsing the file from writing the database. This method is pretty long doing both

austinwiltshire commented on create_database.py:
```diff
@@ -0,0 +1,107 @@
+    """Create an SQLite database file consisting of two tables, 'users'
+    and 'date_formats'.
+    """
+    create_user_table(db_filename, users_filename)
```
This is 'logically' cohesive rather than a higher level. Do you see why? You're basically putting these two function calls together because they both create tables, not because they have any other relation. Why should, for instance, the users filename and the dates formats filename ever need to be passed into the same function? May want to just put this behavior in your main rather than creating a function for it

austinwiltshire commented on find_user.py:
```diff
@@ -0,0 +1,46 @@
+import date_validation
+from parse import parse
+
+def find_user(string, db_filename):
```
take in a database here rather than the file

austinwiltshire commented on find_user.py:
```diff
@@ -0,0 +1,46 @@
+    user = cursor.fetchone()
+    database.close()
+    if user:
+        p_month, p_day, p_year = date_validation.parse_date(user[2])
```
better names than p_... needed

austinwiltshire replied:
may also want to have a 'plain old data' object for date as well, month, day and year are all cohesive concepts and should live together.

you can put your validation in there.

you might want to check out the datetime module, however, you don't have to incorporate it here. if you had to use dates in real code, someone would ask why you're not using datetimes.

austinwiltshire commented on find_user.py:
```diff
@@ -0,0 +1,46 @@
+    database.close()
+    if user:
+        p_month, p_day, p_year = date_validation.parse_date(user[2])
+        username = user[0]
```
so you can either solve this with objects or helper functions, but username = user[0] is not as readable as 

```
user.username()
```

or

```
get_username(user)
```

Getters like these often happen with 'plain old data' coming from a database.

austinwiltshire commented on find_user.py:
```diff
@@ -0,0 +1,46 @@
+        return "Error, user not found."
+
+if __name__ == "__main__":
+    DB_FILENAME = 'database.sqlite' # Insert desired database filename here.
```
let's go ahead and grab this from the command line too

austinwiltshire commented on find_user_by_nationality.py:
```diff
@@ -0,0 +1,42 @@
+import re
+from parse import parse
+
+def find_user_by_nationality(string, db_filename):
```
need a better variable name than just string, and pass in the database object

austinwiltshire commented on find_user_by_nationality.py:
```diff
@@ -0,0 +1,42 @@
+            user_list.append(user[0])
+        return user_list
+    else:
+        return ["Error, no users of that nationality were found."]
```
nice error handling

austinwiltshire commented on find_user_by_nationality.py:
```diff
@@ -0,0 +1,42 @@
+    users = cursor.fetchall()
+    database.close()
+    if users:
+        user_list = []
```
use list comprehensions for this

return [user[0] for user in user_list]

austinwiltshire commented on find_user_by_nationality.py:
```diff
@@ -0,0 +1,42 @@
+        return ["Error, no users of that nationality were found."]
+
+if __name__ == "__main__":
+    DB_FILENAME = 'database.sqlite' # Insert desired database filename here.
```
take this in at command line

austinwiltshire commented on test_find_user.py:
```diff
@@ -0,0 +1,37 @@
+nationality: %s
+birthdate: %s"""
+    # Check mm/dd/yyyy style user.
+    assert find_user('coit125', db_filename) == template % ('coit125',
```
i like the use of templating here

austinwiltshire commented:
Great grasp of SQL. And you're right, other than a few plain old data objects, I didn't see a large argument for more introduction of OO.

You might want to also introduce a database helper with a query method on it, similar to the query helper you have in your tests, that would create a database from a filename, manage cursors for you, and do any other conversion back and forth from the database.

Good work!

austinwiltshire commented:
Also good answer to the question.

CoitThomas committed:
Uncouple file reading and database opening from table creation in create_database module.
Create get_data() helper function for create_database module.
Uncouple database opening & closing from find_user function.
Fix variable type bug in chunk_data() function.
Make find_user module take in the db_filename from the command line.
Create get module which contains getter functions that are shared by the find_user module and find_user_by_nationality module.
Refactor find_user_by_nationality to utilize a list comprehension.
Uncouple database opening & closing from find_user_by_nationality function.
Refactor and add new unit tests for new functions that have no coverage.
Integrate parse.py into create_database.py.
Fix unit test for parse.py.

CoitThomas committed:
Break apart the get module and incorporate its functionality in a more cohesive manner.
Create username, nationality, date_format, and database modules.
Reinstate the parse module.
Convert the date_validation module into the birthdate module.
Add query and query_all functions.
Create unpack_user_input helper functions in the find_user and find_users_by_nationality modules.
Modify, remove, and add unit tests as needed.

CoitThomas committed:
Implement keyword argument for chunk_size in get_data_from_file function.
Uncouple error message from find_users_by_nationality function.
Move file reading/data retrieval out of the create_database.py module and into its own module 'read_file.py'.
Revert variable names with 'file' in them back to a true representation of their content with 'filename'.
Move filename variables from the very beginning of functions to locations more relavant to their usage in the code.
Combine unpack_user_input test modules into one unit test 'test_unpack_user_input.py'.
Move the unpackaging of data outside of the create_X_table functions.
Create is_valid() functions for username and nationality modules.
Create distinction between get() and get_from_label() functions.
Update unit tests accordingly.

CoitThomas committed:
Simplify the logic on functions that return None.
Reflect the true boolean nature of is_valid() functions.
